File: mobile/lib/models/visit_mapping.dart
/// Which catalogue product each (visit count × day-type) combination maps to
/// (Settings > Bookings > Visits). Null slots are unconfigured. Used to create
/// Visit bookings and to tell a Visit entry apart from an ordinary Walk.
class VisitMapping {
  final String? oneVisitWeekday;
  final String? oneVisitWeekend;
  final String? oneVisitBankHoliday;
  final String? twoVisitWeekday;
  final String? twoVisitWeekend;
  final String? twoVisitBankHoliday;
  // Day Care / Boarding product slots (Settings > Bookings), used to tell those
  // day bookings apart from Visits/Walks on the calendar.
  final String? dayCareHalfDay;
  final String? dayCareFullDay;
  final String? dayCareSecondDogHalfDay;
  final String? dayCareSecondDogFullDay;
  final String? boardingPerDay;
  final String? boardingSecondDogPerDay;

  VisitMapping({
    this.oneVisitWeekday,
    this.oneVisitWeekend,
    this.oneVisitBankHoliday,
    this.twoVisitWeekday,
    this.twoVisitWeekend,
    this.twoVisitBankHoliday,
    this.dayCareHalfDay,
    this.dayCareFullDay,
    this.dayCareSecondDogHalfDay,
    this.dayCareSecondDogFullDay,
    this.boardingPerDay,
    this.boardingSecondDogPerDay,
  });

  static String? _id(dynamic v) =>
      v is Map<String, dynamic> ? v['_id'] as String? : (v as String?);

  factory VisitMapping.fromJson(Map<String, dynamic> json) => VisitMapping(
        oneVisitWeekday: _id(json['oneVisitWeekdayProduct']),
        oneVisitWeekend: _id(json['oneVisitWeekendProduct']),
        oneVisitBankHoliday: _id(json['oneVisitBankHolidayProduct']),
        twoVisitWeekday: _id(json['twoVisitWeekdayProduct']),
        twoVisitWeekend: _id(json['twoVisitWeekendProduct']),
        twoVisitBankHoliday: _id(json['twoVisitBankHolidayProduct']),
        dayCareHalfDay: _id(json['dayCareHalfDayProduct']),
        dayCareFullDay: _id(json['dayCareFullDayProduct']),
        dayCareSecondDogHalfDay: _id(json['dayCareSecondDogHalfDayProduct']),
        dayCareSecondDogFullDay: _id(json['dayCareSecondDogFullDayProduct']),
        boardingPerDay: _id(json['boardingPerDayProduct']),
        boardingSecondDogPerDay: _id(json['boardingSecondDogPerDayProduct']),
      );

  Set<String> get _oneVisitIds =>
      {oneVisitWeekday, oneVisitWeekend, oneVisitBankHoliday}.whereType<String>().toSet();
  Set<String> get _twoVisitIds =>
      {twoVisitWeekday, twoVisitWeekend, twoVisitBankHoliday}.whereType<String>().toSet();
  Set<String> get _dayCareIds => {
        dayCareHalfDay,
        dayCareFullDay,
        dayCareSecondDogHalfDay,
        dayCareSecondDogFullDay,
      }.whereType<String>().toSet();
  Set<String> get _boardingIds =>
      {boardingPerDay, boardingSecondDogPerDay}.whereType<String>().toSet();

  /// True if [productId] is any of the six mapped visit products.
  bool isVisitProduct(String productId) =>
      _oneVisitIds.contains(productId) || _twoVisitIds.contains(productId);

  /// True if [productId] is a Day Care product (Half/Full, incl. 2nd-dog rates).
  bool isDayCareProduct(String productId) => _dayCareIds.contains(productId);

  /// True if [productId] is a Boarding product (incl. the 2nd-dog rate).
  bool isBoardingProduct(String productId) => _boardingIds.contains(productId);

  /// 1 or 2 if [productId] is a mapped visit product, else null (a Walk).
  int? visitCountForProduct(String productId) {
    if (_oneVisitIds.contains(productId)) return 1;
    if (_twoVisitIds.contains(productId)) return 2;
    return null;
  }

  /// The mapped product for [visits] (1 or 2) on the given [dayType]
  /// ('weekday' | 'weekend' | 'bank_holiday'), or null if unconfigured.
  String? productFor(int visits, String dayType) {
    if (visits == 1) {
      return switch (dayType) {
        'weekend' => oneVisitWeekend,
        'bank_holiday' => oneVisitBankHoliday,
        _ => oneVisitWeekday,
      };
    }
    return switch (dayType) {
      'weekend' => twoVisitWeekend,
      'bank_holiday' => twoVisitBankHoliday,
      _ => twoVisitWeekday,
    };
  }
}

File: admin/src/utils/visitMapping.ts
import type { VisitMapping } from '../types';

// Which of the 6 Settings > Bookings > Visits product slots a given product
// id belongs to, if any -- used to tell a "Visit" entry (Bookings page's
// Visits section) apart from an ordinary "Walk" entry (anything else).
export function oneVisitProductIds(mapping: VisitMapping): Set<string> {
  return new Set(
    [mapping.oneVisitWeekdayProduct, mapping.oneVisitWeekendProduct, mapping.oneVisitBankHolidayProduct].filter(
      (id): id is string => !!id,
    ),
  );
}

export function twoVisitProductIds(mapping: VisitMapping): Set<string> {
  return new Set(
    [mapping.twoVisitWeekdayProduct, mapping.twoVisitWeekendProduct, mapping.twoVisitBankHolidayProduct].filter(
      (id): id is string => !!id,
    ),
  );
}

export function isVisitProduct(mapping: VisitMapping, productId: string): boolean {
  return oneVisitProductIds(mapping).has(productId) || twoVisitProductIds(mapping).has(productId);
}

/** 1 or 2 if productId is one of the Visits-mapping products, else null (an ordinary Walk product). */
export function visitCountForProduct(mapping: VisitMapping, productId: string): 1 | 2 | null {
  if (oneVisitProductIds(mapping).has(productId)) return 1;
  if (twoVisitProductIds(mapping).has(productId)) return 2;
  return null;
}

export function isDayCareProduct(mapping: VisitMapping, productId: string): boolean {
  return productId === mapping.dayCareHalfDayProduct || productId === mapping.dayCareFullDayProduct;
}

export function isBoardingProduct(mapping: VisitMapping, productId: string): boolean {
  return productId === mapping.boardingPerDayProduct;
}

// AM drop-off + PM collection is the only combination that means the dog is
// there across midday -- anything else (AM/AM, PM/AM, PM/PM) is a shorter
// stay, so Half Day covers it. Returns the configured product id, or null if
// that slot has no product set in Settings > Bookings > Day Care.
export function dayCareProductFor(
  mapping: VisitMapping,
  dropOffPeriod: 'AM' | 'PM',
  collectionPeriod: 'AM' | 'PM',
): string | null {
  const isFullDay = dropOffPeriod === 'AM' && collectionPeriod === 'PM';
  return (isFullDay ? mapping.dayCareFullDayProduct : mapping.dayCareHalfDayProduct) ?? null;
}
